### PR TITLE
[#11572] Notification Feature - GET route

### PR DESCRIPTION
**Pull request by ziqing26**
Part of #11572 

**Outline of Solution**

- [x] Add GET route
- [x] Create `NotificationsData` for API output
- [x] Fetch data from logic layer in `GetNotificationsAction`
- [x] Add Enum for `NotificationType` and `NotificationTargetUser`
- [x] Add validation for parameters

**Things that reviewers might want to take look at**
- GET request flow
- Usage of Enum `NotificationType` and `NotificationTargetUser`

**Not included yet**
- GET API tests (will submit another PR)
- Use isFetchingAll to only fetch unread notifications(Not in Scenario 1 and 2)

### CODE REVIEW

File: src/main/java/teammates/logic/core/NotificationsLogic.java [hunk 1]
```diff
@@ -36,6 +39,15 @@ public NotificationAttributes getNotification(String notificationId) {
         return notificationsDb.getNotification(notificationId);
     }
 
+    /**
+     * Gets a list of notifications.
+     *
+     * @return a list of notifications with the specified {@code notificationIds}.
```
Comment (NicolasCwy) [inline src/main/java/teammates/logic/core/NotificationsLogic.java:45]: Should this be targetUser? Seems like you are querying by that, unless this comment is trying to say something else.

File: src/main/java/teammates/common/datatransfer/NotificationTargetUser.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    public String toSingularFormString() {
+        switch (this) {
+        case STUDENT:
+            return "student";
+        case INSTRUCTOR:
+            return "instructor";
+        case GENERAL:
+            return "general";
+        default:
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/NotificationTargetUser.java:48]: Is a default case necessary for an enum? Assumed it should fit into one of the enums specified.

File: src/main/java/teammates/common/datatransfer/NotificationTargetUser.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    public static NotificationTargetUser getEnum(String targetUser) {
+        switch (targetUser) {
+        case Const.NotificationTargetUser.STUDENT:
+            return STUDENT;
+        case Const.NotificationTargetUser.INSTRUCTOR:
+            return INSTRUCTOR;
+        case Const.NotificationTargetUser.GENERAL:
+            return GENERAL;
+        default:
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/NotificationTargetUser.java:64]: Same as above

File: src/main/java/teammates/common/datatransfer/NotificationType.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        switch (this) {
+        case MAINTENANCE:
+            return "maintenance";
+        case VERSION_NOTE:
+            return "versionnote";
+        case DEPRECATION:
+            return "deprecation";
+        case TIPS:
+            return "tips";
+        default:
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/NotificationType.java:55]: Same as previous enum - is default case necessary?

File: src/main/java/teammates/common/datatransfer/NotificationType.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        case Const.NotificationType.MAINTENANCE:
+            return MAINTENANCE;
+        case Const.NotificationType.VERSION_NOTE:
+            return VERSION_NOTE;
+        case Const.NotificationType.DEPRECATION:
+            return DEPRECATION;
+        case Const.NotificationType.TIPS:
+            return TIPS;
+        default:
+            return null;
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/NotificationType.java:74]: Same as above

File: src/main/java/teammates/common/datatransfer/NotificationTargetUser.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Formats the participant type as a singular noun.
+     *
+     * @return A user-friendly {@code String} representing this participant in singular form.
+     */
+    public String toSingularFormString() {
+        switch (this) {
+        case STUDENT:
+            return "student";
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/NotificationTargetUser.java:43]: Are we able to use `Const.NotificationTargetUser.STUDENT` or is there a reason to hardcode the string? Likewise for the NotificationType enum

File: src/main/java/teammates/common/datatransfer/NotificationType.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    /**
+     * Formats the participant type as a singular noun.
+     *
+     * @return A user-friendly {@code String} representing this participant in singular form.
+     */
+    public String toSingularFormString() {
+        switch (this) {
+        case MAINTENANCE:
+            return "maintenance";
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/NotificationType.java:48]: Query on the hardcoding of `maintenance ` instead of using `Const.NotificationType.MAINTENANCE`

File: src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java [hunk 8]
```diff
@@ -140,6 +144,10 @@ public void setShown() {
     }
 
     public Instant getCreatedAt() {
+        // TODO: temporary fix for null createdAt
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java:147]: Was this for testing? Not sure if createdAt will be null

File: src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java [hunk 9]
```diff
@@ -179,7 +191,7 @@ public void setUpdatedAt(Instant updatedAt) {
     @Override
     public Notification toEntity() {
         return new Notification(notificationId, startTime, endTime, type,
-                targetUser, title, message, shown, createdAt, updatedAt);
+                targetUser, title, message, shown, updatedAt);
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java:194]: I think this is not necessary, see my note in `Notification.java`

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    void checkSpecificAccessControl() throws UnauthorizedAccessException {
+        // Any user can get notifications as long as its parameters are valid
+    }
+
+    @Override
+    public JsonResult execute() throws InvalidHttpRequestBodyException, InvalidOperationException {
+        String targetUser = getRequestParamValue(Const.ParamsNames.NOTIFICATION_TARGET_USER);
```
Comment (NicolasCwy) [inline src/main/java/teammates/ui/webapi/GetNotificationAction.java:30]: Should we have them all be processed in a request? Such as `CourseBasicRequest`, then branch of more specialized requests like `CourseCreateRequest` and `CourseUpdateRequest`

But it feels abit overengineered for a single param value.

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        //        boolean isFetchingAll = Boolean.parseBoolean(
```
Comment (NicolasCwy) [inline src/main/java/teammates/ui/webapi/GetNotificationAction.java:31]: Can remove if not needed anymore

File: src/main/java/teammates/storage/api/NotificationsDb.java [hunk 12]
```diff
@@ -37,6 +42,52 @@ public NotificationAttributes getNotification(String notificationId) {
...
 
+    /**
+     * Gets notifications by {@code targetUser}.
+     *
+     * @return a list of notifications for the specified targetUser.
+     */
+    public List<NotificationAttributes> getNotificationsByTargetUser(NotificationTargetUser targetUser) {
+        assert targetUser != null;
+
+        List<Notification> endEntities = load()
```
Comment (NicolasCwy) [inline src/main/java/teammates/storage/api/NotificationsDb.java:53]: Would we be able to construct an OR filter so that we can filter out targetUser OR "general" at the same time. Not sure if its necessary to do 2 filters and if it will affect performance

Something I saw on [composite filters](https://cloud.google.com/appengine/docs/standard/java/datastore/queries)

File: src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java [hunk 13]
```diff
@@ -165,10 +181,6 @@ public void setUpdatedAt(Instant updatedAt) {
 
         addNonEmptyError(FieldValidator.getInvalidityInfoForTimeForNotificationStartAndEnd(startTime, endTime), errors);
 
-        addNonEmptyError(FieldValidator.getInvalidityInfoForNotificationType(type), errors);
```
Comment (NicolasCwy) [inline src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java:168]: I don't think this can be removed since POST uses this to validate the type

File: src/main/java/teammates/storage/entity/Notification.java [hunk 14]
```diff
@@ -75,19 +78,22 @@ public Notification(Instant startTime, Instant endTime, String type, String targ
      * Instantiates a new notification, with all fields passed in as parameters.
      * This is mainly for conversion from attributes to entity.
      */
-    public Notification(String notificationId, Instant startTime, Instant endTime, String type, String targetUser,
-            String title, String message, boolean shown, Instant createdAt, Instant updatedAt) {
+    public Notification(String notificationId, Instant startTime, Instant endTime,
+                        NotificationType type, NotificationTargetUser targetUser,
```
Comment (NicolasCwy) [inline src/main/java/teammates/storage/entity/Notification.java:82]: Perhaps we can leave this constructor alone? I think this can be used for PUT requests where we would not want createdAt to change, you could use the constructor above for a new notification object

**Review by NicolasCwy — CHANGES_REQUESTED** [review]
Nice work on getting the GET request up, some changes needed and also some points for us to discuss.

File: src/main/java/teammates/common/datatransfer/NotificationTargetUser.java (hunk 6 above)
Comment (zhaojj2209) [reply src/main/java/teammates/common/datatransfer/NotificationTargetUser.java:43]: To take it one step further, would it be possible to simplify the `NotificationTargetUser` and `NotificationTypes` enums to something like `CommentVisibilityType`? It could potentially allow us to remove the respective enums in `Const.java` altogether as well as the functions here, thereby making the code less complicated.

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java (hunk 11 above)
Comment (zhaojj2209) [reply src/main/java/teammates/ui/webapi/GetNotificationAction.java:31]: Seems like this is code for after we start implementing read status? Would be good to add a TODO here

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java (hunk 10 above)
Comment (zhaojj2209) [reply src/main/java/teammates/ui/webapi/GetNotificationAction.java:30]: I'd say the current implementation works fine since there are only two attributes!

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        //                getRequestParamValue(Const.ParamsNames.NOTIFICATION_IS_FETCHING_ALL));
+
+        String targetUserErrorMessage = FieldValidator.getInvalidityInfoForNotificationTargetUser(targetUser);
+        if (!targetUserErrorMessage.isEmpty()) {
+            throw new InvalidHttpRequestBodyException(targetUserErrorMessage);
+        }
+
+        List<NotificationAttributes> notificationAttributes =
+                logic.getNotificationsByTargetUser(NotificationTargetUser.valueOf(targetUser));
```
Comment (zhaojj2209) [inline src/main/java/teammates/ui/webapi/GetNotificationAction.java:40]: Currently, the code only allows retrieving of notifications by target user type; we might need to add extra logic here to retrieve all notifications (either in this PR or future PRs), since it may needed for the admin notifications page.

File: src/main/java/teammates/common/datatransfer/NotificationTargetUser.java (hunk 2 above)
Comment (ziqing26) [reply src/main/java/teammates/common/datatransfer/NotificationTargetUser.java:48]: Without the default case, there is a chance that the method does not return anything from the syntax. So a default statement is needed.

File: src/main/java/teammates/common/datatransfer/NotificationTargetUser.java (hunk 6 above)
Comment (ziqing26) [reply src/main/java/teammates/common/datatransfer/NotificationTargetUser.java:43]: Sure, I think simplify it would be good. Previously I was making use of the enums in `Const.java` as well.

File: src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java (hunk 8 above)
Comment (ziqing26) [reply src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java:147]: Before this PR, the POST request actually creates entities with null createdAt. So this was to view those entities.

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java (hunk 10 above)
Comment (ziqing26) [reply src/main/java/teammates/ui/webapi/GetNotificationAction.java:30]: Agree with Jingjing, and POST and GEt requests for notifications share little similarities, might want to keep it simple first in my opinion.

File: src/main/java/teammates/storage/entity/Notification.java (hunk 14 above)
Comment (ziqing26) [reply src/main/java/teammates/storage/entity/Notification.java:82]: What do you mean by using this for PUT request? As I thought this is for creating a new notification instead of updating certain UpdateOptions

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java (hunk 15 above)
Comment (ziqing26) [reply src/main/java/teammates/ui/webapi/GetNotificationAction.java:41]: Sure, I think this is a good point, I will create another method for getting all notifications

File: src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java (hunk 9 above)
Comment (ziqing26) [reply src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java]: After discussion, we decide to add a conditional check here so that createdAt could be set to `Instant.now()` upon initialization.

File: src/main/java/teammates/storage/api/NotificationsDb.java (hunk 12 above)
Comment (ziqing26) [reply src/main/java/teammates/storage/api/NotificationsDb.java:63]: After searching and trials, I don't think the SDK currently supports `IN` filters. Or if anyone knows a way, please let me know while I search more.

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public JsonResult execute() throws InvalidHttpRequestBodyException, InvalidOperationException {
+        String targetUser = getRequestParamValue(Const.ParamsNames.NOTIFICATION_TARGET_USER);
+        // TODO: Use isFetchingAll to decide whether to fetch unread notification only.
+        // boolean isFetchingAll = Boolean.parseBoolean(
+        //     getRequestParamValue(Const.ParamsNames.NOTIFICATION_IS_FETCHING_ALL));
+
+        List<NotificationAttributes> notificationAttributes;
+        if (targetUser == null & userInfo.isAdmin) {
```
Comment (zhaojj2209) [inline src/main/java/teammates/ui/webapi/GetNotificationAction.java:36]:
```suggestion
        if (targetUser == null && userInfo.isAdmin) {
```

**Review by zhaojj2209 — COMMENTED** [review]
Thanks for the changes! Just one more nit

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java (hunk 16 above)
Comment (ziqing26) [reply src/main/java/teammates/ui/webapi/GetNotificationAction.java]: Thanks!

**Review by NicolasCwy — COMMENTED** [review]
Just some clarifications about invalid NotificationType when the request is deserialized in `CreateNotificationRequest.java`

@jianhandev Would you be able to confirm if that change is ok?

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+            // if the admin wants to retrieve all notifications
+            notificationAttributes =
+                    logic.getAllNotifications();
+        } else {
+            String targetUserErrorMessage = FieldValidator.getInvalidityInfoForNotificationTargetUser(targetUser);
+            if (!targetUserErrorMessage.isEmpty()) {
+                throw new InvalidHttpRequestBodyException(targetUserErrorMessage);
+            }
+            notificationAttributes =
+                    logic.getNotificationsByTargetUser(NotificationTargetUser.valueOf(targetUser));
```
[the remark this answers is not in the document]
Comment (ziqing26) [reply src/main/java/teammates/ui/webapi/GetNotificationAction.java:46]: The types will be generated using `generateType` in `api-output.ts` `api-requests.ts` if applicable, i.e. the generated ENUM values are the same as the name in the frontend, so this should not be an issue.

File: src/main/java/teammates/ui/request/CreateNotificationRequest.java [hunk 18]
```diff
@@ -1,13 +1,16 @@
...
 /**
  * The request for creating a notification.
  */
 public class CreateNotificationRequest extends BasicRequest {
     private Long startTimestamp;
     private Long endTimestamp;
-    private String notificationType;
-    private String targetUser;
+    private NotificationType notificationType;
+    private NotificationTargetUser targetUser;
```
[the remark this answers is not in the document]
Comment (ziqing26) [reply src/main/java/teammates/ui/request/CreateNotificationRequest.java:13]: @t-cheepeng We might need some inputs from senior on how to validate enums. Currently it is validated when we get values for request parameters. If the value does not match any NotificationType, it will be null and will be caught by validation.

File: src/main/java/teammates/ui/webapi/GetNotificationAction.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (jianhandev) [reply src/main/java/teammates/ui/webapi/GetNotificationAction.java:46]: > The types will be generated using `generateType` in `api-output.ts` `api-requests.ts` if applicable, i.e. the generated ENUM values are the same as the name in the frontend, so this should not be an issue.

Yes that is correct.

File: src/main/java/teammates/ui/request/CreateNotificationRequest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (jianhandev) [reply src/main/java/teammates/ui/request/CreateNotificationRequest.java:13]: Your current way of validating is fine, but in `NotificationAttributes` you are going to validate again (for the other fields) before inserting the data into db, so why not do all the validation there instead

**Review by NicolasCwy — APPROVED** [review]
LGTM then!

@jianhandev Thanks for the input

File: src/main/java/teammates/common/util/Const.java [hunk 19]
```diff
@@ -193,6 +173,9 @@ private Const() {
         public static final String QUERY_LOGS_VERSION = "version";
         public static final String QUERY_LOGS_EXTRA_FILTERS = "extrafilters";
         public static final String QUERY_LOGS_ORDER = "order";
+
+        public static final String NOTIFICATION_TARGET_USER = "usertype";
+        public static final String NOTIFICATION_IS_FETCHING_ALL = "isfetchingall";
```
Comment (fsgmhoward) [inline src/main/java/teammates/common/util/Const.java:178]: Document uses camel case. Would you like to make a note at the document regarding the change?

**Review by zhaojj2209 — APPROVED** [review]
LGTM!

File: src/main/java/teammates/storage/api/NotificationsDb.java [hunk 20]
```diff
@@ -37,6 +42,61 @@ public NotificationAttributes getNotification(String notificationId) {
...
+                .filter("targetUser", targetUser)
+                .filter("endTime >", Instant.now())
+                .list();
+
+        List<Notification> startEntities = load()
+                .filter("targetUser", targetUser)
+                .filter("startTime <", Instant.now())
+                .list();
```
Comment (fsgmhoward) [inline src/main/java/teammates/storage/api/NotificationsDb.java:71]: why not use `.filter()` two times to filter the time? i.e.
```
        List<Notification> entities = load()
                .filter("targetUser", targetUser)
                .filter("startTime <", Instant.now())
                .filter("endTime >", Instant.now())
                .list();
```

File: src/main/java/teammates/storage/api/NotificationsDb.java [hunk 21]
```diff
@@ -37,6 +42,61 @@ public NotificationAttributes getNotification(String notificationId) {
...
+
+        if (targetUser != NotificationTargetUser.GENERAL) {
```
Comment (fsgmhoward) [inline src/main/java/teammates/storage/api/NotificationsDb.java:73]: Is this check redundant? I don't think `GENERAL` is an allowed value for `targetUser` parameter from request. It should only be student or instructor (checks should be done at the GET request handler).

File: src/main/java/teammates/storage/api/NotificationsDb.java [hunk 22]
```diff
@@ -37,6 +42,61 @@ public NotificationAttributes getNotification(String notificationId) {
...
+        List<String> startEntitiesIds = startEntities.stream()
+                .map(Notification::getNotificationId)
+                .collect(Collectors.toList());
+
+        List<Notification> ongoingNotifications = endEntities.stream()
+                .filter(notif -> {
+                    String id = notif.getNotificationId();
+                    return startEntitiesIds.contains(id);
+                })
+                .collect(Collectors.toList());
```
Comment (fsgmhoward) [inline src/main/java/teammates/storage/api/NotificationsDb.java:93]: Same comment as above. Would it be better to use two filter methods so we don't have to do these additional processing?

**Review by fsgmhoward — COMMENTED** [review]
LGTM. Just some comments on the code to consider.

File: src/main/java/teammates/common/util/Const.java (hunk 19 above)
Comment (ziqing26) [reply src/main/java/teammates/common/util/Const.java:178]: Okay sure, let me edit the document for this. Thanks for the note!

File: src/main/java/teammates/storage/api/NotificationsDb.java (hunk 20 above)
Comment (ziqing26) [reply src/main/java/teammates/storage/api/NotificationsDb.java:71]: Datastore does not allow queries that uses two attributes inequality. Another example is `getAllOngoingSessions` in `FeedbackSessionsDb.java`.

Comment (fsgmhoward) [reply src/main/java/teammates/storage/api/NotificationsDb.java:71]: You made a point. Would it be ok if you add one composite index for this? I think it works after adding composite indices.

Because if you do query for two times, the cost will be many times higher especially when you have a large number of entities at a later stage. Even there is only one active notification, you will fetch almost everything from the database.

Comment (ziqing26) [reply src/main/java/teammates/storage/api/NotificationsDb.java:71]: Okay let me try that!

Comment (ziqing26) [reply src/main/java/teammates/storage/api/NotificationsDb.java]: Although multiple inequality filter is not supported for composite index, I have removed the query by `startTime` and use Java filter instead as suggested.

**Review by fsgmhoward — APPROVED** [review]
LGTM.

File: src/main/java/teammates/storage/entity/Notification.java [hunk 23]
```diff
@@ -75,19 +78,25 @@ public Notification(Instant startTime, Instant endTime, String type, String targ
...
+                        String title, String message, boolean shown, Instant createdAt, Instant updatedAt) {
         this.setStartTime(startTime);
         this.setEndTime(endTime);
         this.setType(type);
         this.setTargetUser(targetUser);
         this.setTitle(title);
         this.setMessage(message);
-        this.setCreatedAt(createdAt);
+        if (createdAt == null) {
```
Comment (t-cheepeng) [inline src/main/java/teammates/storage/entity/Notification.java:90]: Can just shorten this to 
`this.createdAt = createdAt == null ? this.setCreatedAt(Instant.now()): this.setCreatedAt(createdAt)`

Comment (t-cheepeng) [inline src/main/java/teammates/storage/entity/Notification.java:90]: Can just shorten this to
this.createdAt = createdAt == null ? this.setCreatedAt(Instant.now()): this.setCreatedAt(createdAt)

**Review by t-cheepeng — APPROVED** [review]
Minor code changes. LGTM

**Review by jianhandev — CHANGES_REQUESTED** [review]
Good work! Some areas you can improve on:

File: src/main/java/teammates/storage/entity/Notification.java (hunk 23 above)
Comment (ziqing26) [reply src/main/java/teammates/storage/entity/Notification.java:90]: Because we are using setter here, so we cannot assign a setter to `this.createdAt` unless we do not use a setter.

File: src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java [hunk 24]
```diff
@@ -165,14 +177,12 @@ public void setUpdatedAt(Instant updatedAt) {
...
-
-        addNonEmptyError(FieldValidator.getInvalidityInfoForNotificationTargetUser(targetUser), errors);
-
```
[the remark this answers is not in the document]
Comment (ziqing26) [reply src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java:171]: Previously I thought using ` assert type != null;` and `assert targetUser != null;` in this method would be sufficient. When the attribute is created, the enum is created, it must be either null or one of the value that is in the enum.

[the remark this answers is not in the document]
Comment (jianhandev) [reply src/main/java/teammates/common/datatransfer/attributes/NotificationAttributes.java]: Yes I get what you mean, I guess you are not wrong... but for now let's just keep the validations before db insertion in one place.

**Review by jianhandev — APPROVED** [review]
LGTM!
